### DISCUSSION TITLE
JPA에서 양방향 연관관계가 필요한 경우와, 사용할 때 주의할 점

### DISCUSSION THREAD

**Discussion by nayonsoso** [2024-woowacourse-study/level-interview#131] [Spring Data]
.

**Comment by J-I-H-O** [comment]
**<양방향 연관관계가 필요한 경우>**

DB 관점에서 외래 키는 한쪽 테이블에만 두면 되므로, 단방향 연관관계 만으로도 관계를 표현할 수 있습니다.
양방향 연관관계가 필요한 경우는 다음과 같습니다.

- **데이터 접근 편의성**
    양쪽 엔티티에서 서로를 참조하도록 만들어 편리하게 접근 가능하도록 할 수 있습니다.
    
- **비즈니스 로직의 요구사항**
    요구사항에 따라 양방향 참조가 필요할 수 있습니다.
    예를 들어 Team 엔티티와 Member 엔티티가 존재할 때, 특정 팀에 대한 모든 멤버를 조회하는 로직과, 특정 멤버가 속한 팀을 조회하는 로직이 모두 필요한 경우가 있습니다.
    
- **효율적인 쿼리 작성**
    양방향 연관관계를 사용하면 JPQL을 사용할 때 더 직관적이고 효율적인 쿼리를 작성할 수 있습니다.
    부모 테이블에 매핑되는 엔티티에서 자식 테이블에 매핑되는 엔티티에 대한 연관관계를 정의하지 않으면 자식 엔티티 객체에서는 부모 엔티티 객체의 정보를 얻을 수 있습니다. 
    하지만 부모 엔티티 객체에서는 자식 엔티티 객체에 대한 정보를 얻으려면 자식 엔티티를 모두 조회한 다음, 부모 엔티티 객체와 연관된 객체들만 필터링 하는 방식을 사용해야 합니다.
    

**<양방향 연관관계 사용 시 주의점>**

- **연관관계의 주인 설정**
    외래키를 관리하는 테이블에 매핑되는 엔티티를 연관관계의 주인으로 설정해야 합니다. 
    만약 외래키가 없는 쪽을 연관관계의 주인으로 설정하게 되면 엔티티 객체의 속성을 변경할 때 해당 엔티티와 매핑되는 테이블이 아닌, 다른 테이블의 외래키를 관리하게 되는 꼴이 됩니다.
    
- **연관관계의 주인이 아닌 쪽은 읽기만 가능**
    주인이 아닌 쪽에서 연관관계를 맺고 있는 엔티티를 수정하거나 삭제하더라도 DB에는 영향을 미치지 않습니다.
    예를 들어 Team 엔티티와 Member 엔티티가 양방향 연관관계를 맺고있을 때, Member 가 연관관계의 주인이라면 Team 엔티티에서 member 필드를 수정하더라도 DB에 반영되지 않습니다.
    
- **데이터 일관성 유지**
    데이터의 일관성을 유지하기 위해 두 엔티티 간의 연관관계를 설정할 때 양쪽 모두에 값을 설정해주어야 합니다. 예를 들어 Member 엔티티에 Team 객체를 설정할 때, Team 객체의 members 컬렉션에도 Member 객체를 추가해야 합니다.
    
- **무한 루프에 주의해야 함**
    양방향 연관관계에서는 두 엔티티가 서로를 참조하기 때문에 직렬화 시 무한 루프가 발생할 수 있습니다.
    따라서 직렬화 시 `@JsonIgnore` 어노테이션 등을 사용해 순환 참조를 끊어주어야 합니다.
    같은 맥락에서 toString() 을 사용할 때에도 두 객체의 순환 참조를 끊어주어야 합니다.

https://colabear754.tistory.com/142

**Comment by le2sky** [comment]
객체 그래프 탐색에 양방향이 필요한 경우 양방향 매핑을 고려해볼 수 있습니다.
자바 객체인 Member와 Reservation에서 양방향 데이터 탐색이 불가능합니다.
객체 간 양방향 탐색을 가능하게 하려면 상대 객체를 참조하는 필드를 양측이 가져야합니다.

양방향 매핑에서 다음과 같은 부분을 주의해야합니다.

1.  연관관계의 주인을 설정하는 것에 주의해야합니다. 다대일 양방향의 경우 연관관계 주인은 다쪽인 Reservation.member 입니다. 이때 주인이 아닌 Member.reservations는 조회를 위한 JPQL이나 객체 그래프 탐색할 때 사용합니다.

2. 양방향 연관관계를 설정하고 연관관계의 주인에 값을 입력하지 않고, 주인이 아닌 곳에만 값을 입력하는 것을 유의해야합니다. 연관관계의 주인만이 외래 키의 값을 변경할 수 있습니다.

3. 연관관계의 주인에만 값을 저장하고, 주인이 아닌 곳에는 값을 저장하지 않는 경우 문제가 발생할 수 있습니다.
객체 관점에서 양쪽 방향에 모두 값을 입력해주는 것이 가장 안전합니다. 이럴때 연관관계 편의 메서드를 만들어서 사용할 수 있습니다. 

4. 양방향 매핑에서는 두 객체가 서로를 참조하기 때문에 무한 루프 문제에 유의해야합니다.
무한 루프 문제는 상황에 따라`@JsonIgnore`, `@ToString.Exclude`와 같은 어노테이션을 사용하여 예방할 수 있습니다.
